introduces offset for limit filter

tobibeer opened the pull request:
allows to shift the considered tiddlers by a given number, e.g. for paging

aelocson commented:
Another possible place for this extension would be on the `nth` operator, such that `[nth:20[61]]` would select items 61 to 80. Would that be clearer, I wonder?

tobibeer commented:
> Another possible place for this extension would be on the nth operator, such that [nth:20[61]] would select items 61 to 80. Would that be clearer, I wonder?

Why would you think of it as clearer? You sill need to know that the means "starting from". Perhaps both could be aliases... but then I am not fan of those as stated elsewhere.

aelocson commented:
I don't know for sure whether it would be clearer. It seems 50:50 to me. There are two parameters (the startpoint and the count). The `nth` operator currently lets you specify the startpoint but not the count. The `limit` operator currently lets you specify the count but not the startpoint. The symmetry is striking. The question is: which of the operators deserves to be extended to accept the other parameter?

Jermolene commented:
I'm not sure that adding a suffix to the limit operator is the best way to go for this. We could instead introduce a "pop" operator that removes the top 'n' entries from the list. So `[pop[61]limit[20]]` would select items 61 to 80 (zero indexed). "pop" isn't maybe the best name: "discard", "remove", "trim" might be other options. The attraction of using two separate operators is that one can then use transclusion for the operand.

tobibeer commented:
I'm on board with an extra filter, if needed.  I like **pop**, but — like **discard**, **ignore**, **remove** — it does not seem clearer than **offset**. As for **trim**, that notion strikes my mind as cutting off empty strings.

**offset** is a term ubiquitously used for precisely doing this in db queries.

Curiously, the negation **[!offset[20]]** should be **[limit[20]]**. So, in fact, we would already have a working offset if the negation of **limit** wasn't what it currently is!

Therefore, considering #1420, I would strongly advise to always have a negation form the complement to the operation with respect to source. With the current **limit** filter, this is not what happens and I think it's basically wrong to do so. If I wanted to get results in reverse, I reverse the **sort**, not negate the **limit**.

If I wanted to have the last x items, I'd rather use a negative operand for limit, e.g. **[limit[-20]]**.

The limit filter also somewhat inefficiently first adds all source tiddlers to the results array and only later splices things off. I'd think that's more costly than doing some counting and only writing what's necessary.

Why combine **offset** with **limit**? Because: efficiency. Why would I first cut off 20, doing some more superfluous iterating and then do much the same with limit?

The number one reason why not to combine the two for me is that I cannot use evaluated variables or text-references in suffixes at the moment, can I? If not, I think that would be helpful.

aelocson commented:
> If I wanted to have the last x items, I'd rather use a negative operand for limit, e.g. [limit[-20]].

Excellent point.

Jermolene commented:
> With the current limit filter, this is not what happens and I think it's basically wrong to do so. If I wanted to get results in reverse, I reverse the sort, not negate the limit.

Changing the behaviour of negation with the "limit" filter operator sounds like an unacceptably incompatible change, surely?

tobibeer commented:
I think it is not unacceptably incompatible in the sense that if I changed an implementation that does `1+1=3` to finally do `1+1=2` wouldn't be.

Jermolene commented:
The current docs state that `!limit[n]` is a synonym for `last[n]`. So the current behaviour is documented and is the subject of a test. It's easy to disagree with the way that it works, but that is not (on its own) enough to allow us to disregard our commitment to backwards compatibility.

tobibeer commented:
> The current docs state that !limit[n] is a synonym for last[n]. So the current behaviour is documented and is the subject of a test. 

I know. Nevertheless, there are notions that come with a standard meaning. For one, I think it is important to respect those and, secondly, adapting a thus negated limit to a new, more coherent definition shouldn't be something that will majorly break anything. So, I think it is good, even vital not to shy away from indeed *fixing* this.

aelocson commented:
We could perhaps introduce a *new* operator whose behaviour was what you want for `limit`.

tobibeer commented:
> We could perhaps introduce a new operator whose behaviour was what you want for limit.

I have thought about that, too. But I think there are good reasons not to do that. The first being that this is not a big break. A more important reason being that there would still be a filter inconsistently implementing a negation that essentially isn't one, defining it's own brand of set theory so to speak. We can sure pull this oddity through the ages, if you feel like backwards compatibility has highest priority, even if we're compatible with ... you can fill-in the assessment yourself.

We could implement an offset filter that works quite complementary to the initial pull about extending limit, only vice-versa...

`[offset:<limit>[<offset>]]`

namely...

* `[offset[10]]`
    * remove first ten

* `[offset[-10]]`
    * remove last ten

* `[offset:10[10]]`
    * remove first 10, **then** limit to 10

* `[offset:10[-10]]`
    * remove last 10, **then** limit to remaining last 10

The thing is, only creating an offset filter without a limit suffix or vice versa still leaves us with the fact that **limit** is unusable in terms of using its proper negation.


Anyhow, these two variants don't seem to make much sense, but maybe they do...

* `[offset:-10[10]]`
    * remove first 10, then limit to last 10, surely a bit weird a combination

* `[offset:10[-10]]`
    * remove last 10, then limit to first 10, another possibly weird combination

Jermolene commented:
What's the current proposal, exactly?

tobibeer commented:
Give me a second...

tobibeer commented:
I prefer both limit and offset combined as proposed and pushed an update reflecting that.

**Demo:** [limit:offset](http://tbdemo.tiddlyspot.com/#limit%3Aoffset)

**Rationale**

Let's fix and extend **limit** since...

1. currently, the negation of **limit** is...
    1. not conforming to set theory
    2. impinging on practicality — we can't construct a proper negation
2. combining it with an offset suffix is more efficient in terms of processing
    * compared to having separate filters
3. currently, this modification should hardly "break" mission critical implementations

Of course, I'd leave it to you to make that decision.

**Summary**

* modified
    * **$:/core/modules/filters/limit.js**
* introduces
    * offset suffix
    * negative limit and offset (from end)
* changed
    * negation of limit is now proper negation
    * negative limit takes last x
    * positive offset removes first x
    * negative offset removes last x
    * first offset applied, then limit

Jermolene commented:
OK so the remaining issue (after coding conventions) is whether we can change the behaviour of `!limit[n]` in this way. We all agree that the new way is more logical, the question is whether we're going to break anything in the wild.

`!limit[n]` isn't used by the core (there is a test for it though). It's hard to tell whether it's use is common in the field. (Parenthetically, it would be useful sometimes to be able to search all the public TiddlyWiki documents to be able to analyse common usages).

tobibeer commented:
> OK

:+1:  Thanks.

> the question is whether we're going to break anything in the wild

I would leave it to the release notes to indicate this backwards-incompatible change with a short note as to why ...and refer to that if someone would even notice.

> Parenthetically, it would be useful sometimes to be able to search all the public TiddlyWiki documents to be able to analyse common usages

yes

Jermolene commented:
Great, thanks @tobibeer.

I'm afraid we're back to the need for documentation updates to reflect the change :)

aelocson commented:
Let's ask the Google group if anyone's using `!limit`. My guess is that the answer will be "no", simply because of the existence of the much more discoverable synonym `last`.

tobibeer commented:
> Let's ask the Google group if anyone's using !limit. My guess is that the answer will be "no", simply because of the existence of the much more discoverable synonym last.

I am not in favor of this. To me, negation is a global operation that should always be performed when asked. If I set `[!filter[]]`, I always expect to be given the negation in terms of set theory, whether anyone thinks it makes sense or not... it's mathematics.

Jermolene commented on editions/tw5.com/tiddlers/filters/limit.tid:
```diff
@@ -8,12 +8,15 @@ purpose: select the first or last <<.place n>> input titles
 parameter="an integer, defaulting to 0"
 paramName="n"
 output="the first <<.place n>> input titles"
-negationOutput="the last <<.place n>> input titles"
```
The code still seems to process the negation `!`, why aren't we documenting the behaviour?

aelocson commented on editions/tw5.com/tiddlers/filters/limit.tid:
```diff
@@ -8,12 +8,15 @@ purpose: select the first or last <<.place n>> input titles
 
 <$macrocall $name=".operator-def"
 input="a [[selection of titles|Title Selection]]"
+suffix="optional, an offset"
```
For consistency with the documentation of the other filter operators, this should say:

suffix="optionally, an offset, defaulting to 0"

I've added another note below about whether "offset" is the clearest term.

Jermolene commented:
> If I set `[!filter[]]`, I expect to be given the negation in terms of set theory, whether anyone thinks it makes sense or not... it's mathematics.

But that's not how the `!` prefix is defined. Each filter is free to use the negation flag in any way that makes sense. It's too late to try to recast `!` globally as set theory operation.

aelocson commented on editions/tw5.com/tiddlers/filters/limit.tid:
```diff
@@ -8,12 +8,15 @@ purpose: select the first or last <<.place n>> input titles
 
-`limit[n]` and `!limit[n]` are synonyms for `first[n]` and `last[n]` respectively.
+The `limit[n]` filter returns at most the first <<.place n>> items from the input set. Using a negative  <<.place -n>> applies the limit from the end, thus returning the last  <<.place n>> elements.
+
```
The terminology and phrasing here is inconsistent with the revised filter documentation, and the sentence "Using a negative …" effectively says the same thing twice. I suggest this wording:

`limit[n]` outputs the first <<.place n>> input titles. If <<.place n>> is negative, the <<.em last>> <<.place n>> titles are output instead. If the input contains less than <<.place n>> titles, the output is the same as the input.

aelocson commented on editions/tw5.com/tiddlers/filters/limit.tid:
```diff
@@ -8,12 +8,15 @@ purpose: select the first or last <<.place n>> input titles
-`limit[n]` and `!limit[n]` are synonyms for `first[n]` and `last[n]` respectively.
+The `limit[n]` filter returns at most the first <<.place n>> items from the input set. Using a negative  <<.place -n>> applies the limit from the end, thus returning the last  <<.place n>> elements.
+
+You can specify an offset in the suffix. A positive offset <<.place o>> removies the first <<.place o>> elements from the beginning of the input set. A negative offset <<.place -o>> removes the last elements from the input set. The offset is applied before the limit. 
```
Similarly:

The optional suffix specifies the number of titles to trim from the start of the input. If the suffix is negative, titles are trimmed from the <<.em end>> of the input instead. Trimming takes place <<.em before>> the output titles are extracted, thus allowing them to be extracted from somewhere in the middle of the input.

aelocson commented:
Maybe I misunderstood, or missed something in the discussion, but I thought we were going to add a new `discard` operator to trim the start and end of a selection, rather than shoehorning this functionality into the suffix of `limit`.

aelocson commented:
Regarding my proposal to ask the Google group, this was intended as politeness to any people who might be relying on the existing behaviour. I certainly didn't mean to suggest that set theory is a matter of opinion :)

Jermolene commented:
> Maybe I misunderstood, or missed something in the discussion, but I thought we were going to add a new discard operator to trim the start and end of a selection, rather than shoehorning this functionality into the suffix of limit.

Yes, I think you are right, thank you @aelocson, my apologies for any confusion. @tobibeer are you happy for me to close this pull request?

aelocson commented:
**remove** would also work well. I take your point about **trim**.

One possible solution to the negation problem would be to introduce an *alternative* prefix (e.g. `~`) that could behave the way you want without affecting backward compatibility.

tobibeer commented:
Personally, I would not invent or propose any more confusing symbols. I think we've been exhaustively debating "negation". So, I know how to fix this and I will suggest others to do the same, whether the core adapts it or not.

aelocson commented:
I've just realised that we already have operators for discarding items: `butfirst` and `butlast`.

tobibeer commented:
Thanks for pointing out, @aelocson. I find **rest** really confusing to read, e.g. the examples. It's not owed to the examples, but to the naming of the filter. Possibly, **butfirst** and **butlast** also didn't ring that bell with me either, initially.

As that was suggested to be a prime measure, in terms of performance, having two or more filter runs doing what could be done in one also seems less performing.

So, the proposed limit filter would have been good for merging all of these into one...

* `limit[]`
* `butfirst[]` / `bf[]` / `rest[]` and their non-existing negation
* `butlast[]`/ `bl[]`  and their non-existing negation
* that non-existing, proper negation to the limit filter

Jermolene commented:
> As that was suggested to be a prime measure, in terms of performance, having two or more filter runs doing what could be done in one also seems less performing.

The trade-off is interesting. Combining primitive filters gives us more flexibility but I agree that it can affect performance. Bit of a sidenote, but I spent too long early on (2012) investigating compiling the filters to real JS functions. I realised in the end that it was premature optimisation, but it remains our trump card for improving filter performance across the board.

tobibeer commented:
>  I realised in the end that it was premature optimisation, but it remains our trump card for improving filter performance across the board.

Ok, so let's explore that at some later point. Although, I can imagine a "performance sprint" not beeing too far down the road.

Jermolene commented:
> Although, I can imagine a "performance sprint" not being too far down the road.

Yes, we've got some good examples like Richard Smith's Maths textbook that we can use to measure real-world performance.

pmario commented:
>The trade-off is interesting. Combining primitive filters gives us more flexibility but I agree that it can affect performance. Bit of a sidenote, but I spent too long early on (2012) investigating compiling the filters to real JS functions. I realised in the end that it was premature optimisation, but it remains our trump card for improving filter performance across the board.

I did test tiddlywik with the raspberry pi 2 - 4 core ... And tiddlywiki is unusable. 3 seconds to open a tiddler. Every keystroke needs about 1 second. ... It would be nice, if we could add some tests, that tell us about performance regressions. ...

Jermolene commented:
> I did test tiddlywik with the raspberry pi 2 - 4 core ... And tiddlywiki is unusable (in my opinion). 3 seconds to open a tiddler. Every keystroke needs about 1 second.

Yes indeed. I suspect that static renderings are always going to be best for these very constrained environments.

> It would be nice, if we could add some tests, that tell us about performance regressions. ...

Yes, that would be nice, contributions welcome!
